### ISSUE TITLE
Distributed inserts (_bulk, replace, update)

### ISSUE THREAD

**Issue by donhardman** [manticoresoftware/manticoresearch-buddy#392]
### Proposal:

We should implement the logic and stick to JSON protocol ONLY for the initial version when we are able to insert into sharded tables on the Buddy side.

Key considerations:
- Implement `id` generation on the Buddy side that is easy to maintain and can be moved to the daemon part later
- Ensure we use a proper algorithm for sharding logic. If we allow users to pass `id`, we should use MD5 or similar hashing; otherwise, don't allow passing IDs
- For replace or update operations:
  - When performed without an ID, send to all nodes
  - When performed with an ID, we can determine the specific shard from the ID

### Checklist:

<sup>To be completed by the assignee. Check off tasks that have been completed or are not applicable.</sup>
<details>

- [ ] Implementation completed
- [ ] Tests developed
- [ ] Documentation updated
- [ ] Documentation reviewed
- [ ] Changelog updated
- [x] OpenAPI YAML updated and issue created to rebuild clients

</details>

**Comment by donhardman** [comment]
# Two Approaches for ID Generation and Sharding

## 1. Snowflake-like ID Generation
Even distribution is prioritized. This approach generates unique IDs with embedded shard information.

### Structure (63-bit integer):
- 41 bits: timestamp (milliseconds since custom epoch)
- 10 bits: shard/node ID (0-1023)
- 12 bits: sequence number (0-4095)

### Formula:
```php
$id = ($timestamp << 22) | ($shardId % 1024 << 12) | ($sequence);
```

### Example in PHP:
```php
class SnowflakeGenerator {
    private const CUSTOM_EPOCH = 1640995200000; // 2022-01-01
    private $sequence = 0;
    private $lastTimestamp = -1;

    public function generateId($shardId) {
        $timestamp = $this->getCurrentTimestamp();

        if ($timestamp == $this->lastTimestamp) {
            $this->sequence = ($this->sequence + 1) & 4095;
            if ($this->sequence == 0) {
                $timestamp = $this->waitNextMillis($this->lastTimestamp);
            }
        } else {
            $this->sequence = 0;
        }

        $this->lastTimestamp = $timestamp;

        return (($timestamp - self::CUSTOM_EPOCH) << 22)
             | ($shardId % 1024 << 12)
             | $this->sequence;
    }

    // Helper methods...
}
```

### Characteristics:
- Number of shards limited but not fixed
- Cannot support custom IDs
- No extra overhead
- Guaranteed sequential IDs

## 2. MD5-based Sharding
For scenarios where custom IDs are needed.

### Formula:
```php
$shardNumber = hexdec(substr(md5($id), 0, 8)) % $totalShards;
```

### Example in PHP:
```php
class Md5Sharding {
    private $totalShards;

    public function __construct($totalShards) {
        $this->totalShards = $totalShards;
    }

    public function getShardNumber($id) {
        return hexdec(substr(md5((string)$id), 0, 8)) % $this->totalShards;
    }
}

// Usage example
$sharding = new Md5Sharding(16);
$customId = 12345;
$shardNumber = $sharding->getShardNumber($customId);
```

### Characteristics:
- Fixed number of shards
- Supports custom IDs
- Small performance overhead for MD5 calculation
- Even distribution across shards

Both approaches have their use cases:
- Use Snowflake when you need sequential IDs and embedded shard information
- Use MD5-based sharding when you need to support custom IDs or have a fixed number of shards

**Comment by sanikolaev** [comment]
As we discussed in Slack, let's avoid using the snowflake ID approach, as we need to keep the option to provide custom IDs. Instead of the modulo function, let's explore other options, like jump consistent hashing.

**Comment by sanikolaev** [comment]
@donhardman also, as we discussed `select uuid_short()` may be required in the daemon. I've discussed it with @tomatolog and it's not a big deal to add it. Pls create a separate task about it if required.

**Comment by donhardman** [comment]
I have created a task: https://github.com/manticoresoftware/manticoresearch/issues/2752

**Comment by sanikolaev** [comment]
https://github.com/manticoresoftware/manticoresearch/issues/2752 is done

**Comment by donhardman** [comment]
Buddy:

```
Testing with following parameters:
Curl command: curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk
Number of requests: 1000
Concurrency: 12
----------------------------------------
Single-thread curl test (sequential):
----------------------------------------
Progress: 1000/1000
Total time: 63.841850320s
Average time per request: 63.841ms

Concurrent test using parallel execution:
----------------------------------------
Total time: 14.040817747s
Average time per request: 14.040ms
```

Direct bulk insert into table:

```
Testing with following parameters:
Curl command: curl -H 'Content-type: application/x-ndjson' --data-binary @bulk-direct.json http://localhost:9308/_bulk
Number of requests: 1000
Concurrency: 12
----------------------------------------
Single-thread curl test (sequential):
----------------------------------------
Progress: 1000/1000
Total time: 7.573245040s
Average time per request: 7.573ms

Concurrent test using parallel execution:
----------------------------------------
Total time: 1.792843179s
Average time per request: 1.792ms
```

The bulk file has 1024 rows with single document.


<details>
<summary>Script used to test it</summary>


```bash
#!/bin/bash

# Check if required parameters are provided
if [ "$#" -lt 3 ]; then
	echo "Usage: $0 \"<curl_command>\" <number_of_requests> <concurrency>"
	echo "Example: $0 \"curl -X GET http://example.com\" 100 2"
	exit 1
fi

# Store parameters
CURL_CMD="$1"
NUM_REQUESTS=$2
CONCURRENCY=$3

echo "Testing with following parameters:"
echo "Curl command: $CURL_CMD"
echo "Number of requests: $NUM_REQUESTS"
echo "Concurrency: $CONCURRENCY"
echo "----------------------------------------"

echo "Single-thread curl test (sequential):"
echo "----------------------------------------"

# Run sequential curl requests and measure time
start_time=$(date +%s.%N)
for ((i=1; i<=$NUM_REQUESTS; i++)); do
	eval "$CURL_CMD" >/dev/null 2>&1
	echo -ne "\rProgress: $i/$NUM_REQUESTS"
done
echo
end_time=$(date +%s.%N)

# Calculate average time for sequential requests
duration=$(echo "$end_time - $start_time" | bc)
avg_time=$(echo "scale=3; $duration * 1000 / $NUM_REQUESTS" | bc)

echo "Total time: ${duration}s"
echo "Average time per request: ${avg_time}ms"
echo ""

echo "Concurrent test using parallel execution:"
echo "----------------------------------------"

# Create a temporary file to store the commands
temp_file=$(mktemp)
for ((i=1; i<=$NUM_REQUESTS; i++)); do
	echo "$CURL_CMD >/dev/null 2>&1" >> "$temp_file"
done

# Run commands concurrently using parallel
start_time=$(date +%s.%N)
parallel -j "$CONCURRENCY" < "$temp_file"
end_time=$(date +%s.%N)

# Calculate average time for concurrent requests
duration=$(echo "$end_time - $start_time" | bc)
avg_time=$(echo "scale=3; $duration * 1000 / $NUM_REQUESTS" | bc)

echo "Total time: ${duration}s"
echo "Average time per request: ${avg_time}ms"

# Cleanup
rm "$temp_file"
```
</details>

**Comment by donhardman** [comment]
While we are waiting for things that are blocking our next move, let's cover the functionality with tests that we can already implement and manually verify.

Location: https://github.com/manticoresoftware/manticoresearch/pull/2784/

Documentation reference for test cases: https://manual.manticoresearch.com/dev/Data_creation_and_modification/Adding_documents_to_a_table/Adding_documents_to_a_real-time_table?client=Elasticsearch#Auto-ID

We should test the following Elasticsearch-like endpoints with distributed tables:
1. `insert`
2. `replace`
3. `update`
4. `_bulk`
5. `delete`

Additionally, we need to verify that these operations work with distributed tables that have remote agents. We can use sharding to create such table configurations for testing.

**Comment by sanikolaev** [comment]
Here's how the new functionality can be tested and one bug:

```
snikolaev@dev2:~$ docker run --name kit --rm -it ghcr.io/manticoresoftware/manticoresearch:test-kit-buddy-402 bash

root@9356d020d256:/# searchd
Manticore 6.3.9 9183ab762@24112704 dev (columnar 2.3.1 edadc69@24112219) (secondary 2.3.1 edadc69@24112219) (knn 2.3.1 edadc69@24112219)
...

root@9356d020d256:/# mysql -P9306 -h0
mysql> create table t(f text) shards=3;

root@9356d020d256:/# curl 0:9308/insert -d '{"table": "t", "doc": {"f": "abc"}}'
{"error":{"type":"parse_exception","reason":"Document ids should be integer or array of integers","table":"t_s2","index":"t"},"status":400}
```

Same against a non-distributed table works fine:
```
root@9356d020d256:/# curl 0:9308/insert -d '{"table": "l", "doc": {"f": "abc"}}'
{"table":"l","_id":6776366998629646347,"created":true,"result":"created","status":201}
```

**Comment by sanikolaev** [comment]
One more bug: all docs are routed to the same shard:

```
root@9356d020d256:/# mysql -P9306 -h0 -e "create table t(f text) shards=3 rf=1;"

root@9356d020d256:/# for n in `seq 1 200`; do curl -s 0:9308/insert -d '{"table": "t", "id": 0, "doc": {"f": "abc"}}' > /dev/null; done;
root@9356d020d256:/# mysql -P9306 -h0 -e "select count(*) from t"
+----------+
| count(*) |
+----------+
|      200 |
+----------+
root@9356d020d256:/# mysql -P9306 -h0 -e "select count(*) from t_s0"
+----------+
| count(*) |
+----------+
|      200 |
+----------+
root@9356d020d256:/# mysql -P9306 -h0 -e "select count(*) from t_s1"
+----------+
| count(*) |
+----------+
|        0 |
+----------+
```

**Comment by donhardman** [comment]
For make it simpler to write tests here what I used to test:

```
curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk
```

Same for `/insert` and other endpoitns.

And here what I used for example of document

<details>
<summary>insert, update, replace</summary>

```json
{
  "index": "a",
  "id": 1,
  "doc":
  {
    "value": "Hello world"
  }
}
```

</details>

<details>
<summary>delete</summary>

```json
{
  "index": "test",
  "id": 1
}

```
</details>

<details>
<summary>bulk</summary>

```json
{"index":{"_index":"test"}}
{"value":"Donald John Trump (born June 14, 1946) is an American politician, media personality, and businessman who served as the 45th president of the United States from 2017 to 2021. He won the 2024 presidential election as the nominee of the Republican Party and is scheduled to be inaugurated as the 47th president on January 20, 2025.\n\n Trump graduated with a bachelor's degree in economics from the University of Pennsylvania in 1968. After becoming president of the family real estate business in 1971, he renamed it the Trump Organization. After a series of bankruptcies in the 1990s, he launched side ventures, mostly licensing the Trump name. From 2004 to 2015, he produced and hosted the reality television series The Apprentice. In 2015 he launched a presidential campaign.\n\n Trump won the 2016 presidential election. His election and policies sparked numerous protests. In his first term, he ordered a travel ban targeting Muslims and refugees, funded the Trump wall expanding the U.S.–Mexico border wall, and implemented a family separation policy at the border. He rolled back more than 100 environmental policies and regulations, signed the Tax Cuts and Jobs Act of 2017,[a] and appointed three justices to the Supreme Court.[b] He initiated a trade war with China, withdrew the U.S. from several international agreements,[c] and met with North Korean leader Kim Jong Un without progress on denuclearization. He responded to the COVID-19 pandemic with the CARES Act,[d] and downplayed its severity. He was impeached in 2019 for abuse of power and obstruction of Congress, and in 2021 for incitement of insurrection; the Senate acquitted him in both cases."}
```
</details>

**Comment by donhardman** [comment]
The latest test shows the following benchmarks on my dev machine

2 shards
```
root@dev:/workdir# ./test.sh "curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk" 1000 12               
Testing with following parameters:
Curl command: curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk
Number of requests: 1000
Concurrency: 12
----------------------------------------
Single-thread curl test (sequential):
----------------------------------------
Progress: 1000/1000
Total time: 41.567602409s
Average time per request: 41.567ms

Concurrent test using parallel execution:
----------------------------------------
Total time: 11.546097536s
Average time per request: 11.546ms
```

4 shards:

```
root@dev:/workdir# ./test.sh "curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk" 1000 12
Testing with following parameters:
Curl command: curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk
Number of requests: 1000
Concurrency: 12
----------------------------------------
Single-thread curl test (sequential):
----------------------------------------
Progress: 1000/1000
Total time: 33.434813981s
Average time per request: 33.434ms

Concurrent test using parallel execution:
----------------------------------------
Total time: 11.442212978s
Average time per request: 11.442ms
```

8 shards:

```
root@dev:/workdir# ./test.sh "curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk" 1000 12
Testing with following parameters:
Curl command: curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk
Number of requests: 1000
Concurrency: 12
----------------------------------------
Single-thread curl test (sequential):
----------------------------------------
Progress: 1000/1000
Total time: 28.625276014s
Average time per request: 28.625ms

Concurrent test using parallel execution:
----------------------------------------
Total time: 13.499586667s
Average time per request: 13.499ms
```

**Comment by PavelShilin89** [comment]
@donhardman This issue blocked https://github.com/manticoresoftware/manticoresearch/issues/2963

**Comment by donhardman** [comment]
SQL load bench

Command used with shards changed:

```
manticore-load --batch-size=500 --threads=1 --total=500000 --init="drop sharded table test; drop table if exists test; create table test(id bigint, value text) shards=8 rf=1" --load="insert into test(value) values('Donald John Trump (born June 14, 1946) is an American politician, media personality, and businessman who served as the 45th president of the United States from 2017 to 2021. He won the 2024 presidential election as the nominee of the Republican Party and is scheduled to be inaugurated as the 47th president on January 20, 2025. Trump graduated with a bachelors degree in economics from the University of Pennsylvania in 1968. After becoming president of the family real estate business in 1971, he renamed it the Trump Organization. After a series of bankruptcies in the 1990s, he launched side ventures, mostly licensing the Trump name. From 2004 to 2015, he produced and hosted the reality television series The Apprentice. In 2015 he launched a presidential campaign. Trump won the 2016 presidential election. His election and policies sparked numerous protests. In his first term, he ordered a travel ban targeting Muslims and refugees, funded the Trump wall expanding the U.S.–Mexico border wall, and implemented a family separation policy at the border. He rolled back more than 100 environmental policies and regulations, signed the Tax Cuts and Jobs Act of 2017,[a] and appointed three justices to the Supreme Court.[b] He initiated a trade war with China, withdrew the U.S. from several international agreements,[c] and met with North Korean leader Kim Jong Un without progress on denuclearization. He responded to the COVID-19 pandemic with the CARES Act,[d] and downplayed its severity. He was impeached in 2019 for abuse of power and obstruction of Congress, and in 2021 for incitement of insurrection; the Senate acquitted him in both cases.')" --quiet
```

Direct:
```
Threads ; Batch     ; Time        ; Total Docs  ; Docs/Sec    ; Avg QPS     ; p99 QPS     ; p95 QPS     ; p5 QPS      ; p1 QPS      ; Lat Avg     ; Lat p50     ; Lat p95     ; Lat p99     ;
1       ; 500       ; 00:49       ; 500000      ; 10142       ; 20          ; 22          ; 22          ; 18          ; 18          ; 49.3        ; 51.5        ; 57.5        ; 59.5        ;
```

2 shards:
```
Threads ; Batch     ; Time        ; Total Docs  ; Docs/Sec    ; Avg QPS     ; p99 QPS     ; p95 QPS     ; p5 QPS      ; p1 QPS      ; Lat Avg     ; Lat p50     ; Lat p95     ; Lat p99     ;
1       ; 500       ; 00:47       ; 500000      ; 10543       ; 21          ; 23          ; 22          ; 20          ; 19          ; 47.4        ; 47.5        ; 52.5        ; 56.5        ;
```

4 shards:
```
Threads ; Batch     ; Time        ; Total Docs  ; Docs/Sec    ; Avg QPS     ; p99 QPS     ; p95 QPS     ; p5 QPS      ; p1 QPS      ; Lat Avg     ; Lat p50     ; Lat p95     ; Lat p99     ;
1       ; 500       ; 00:35       ; 500000      ; 14069       ; 28          ; 30          ; 30          ; 25          ; 25          ; 35.5        ; 35.5        ; 39.5        ; 50.5        ;
```

8 shards:
```
Threads ; Batch     ; Time        ; Total Docs  ; Docs/Sec    ; Avg QPS     ; p99 QPS     ; p95 QPS     ; p5 QPS      ; p1 QPS      ; Lat Avg     ; Lat p50     ; Lat p95     ; Lat p99     ;
1       ; 500       ; 00:33       ; 500000      ; 14994       ; 30          ; 33          ; 32          ; 27          ; 21          ; 33.3        ; 32.5        ; 39.5        ; 49.5        ;
```

**Comment by sanikolaev** [comment]
Let's compare that with manual sharding, e.g. here we do the same like with `shards=2`, but by writing to the shards manually:

```
manticore-load --batch-size=500 --threads=1 --total=500000 --quiet --init="drop table if exists test; create table test(id bigint, value text)" --load="insert into test(value) values('Donald John Trump (born June 14, 1946) is an American politician, media personality, and businessman who served as the 45th president of the United States from 2017 to 2021. He won the 2024 presidential election as the nominee of the Republican Party and is scheduled to be inaugurated as the 47th president on January 20, 2025. Trump graduated with a bachelors degree in economics from the University of Pennsylvania in 1968. After becoming president of the family real estate business in 1971, he renamed it the Trump Organization. After a series of bankruptcies in the 1990s, he launched side ventures, mostly licensing the Trump name. From 2004 to 2015, he produced and hosted the reality television series The Apprentice. In 2015 he launched a presidential campaign. Trump won the 2016 presidential election. His election and policies sparked numerous protests. In his first term, he ordered a travel ban targeting Muslims and refugees, funded the Trump wall expanding the U.S.–Mexico border wall, and implemented a family separation policy at the border. He rolled back more than 100 environmental policies and regulations, signed the Tax Cuts and Jobs Act of 2017,[a] and appointed three justices to the Supreme Court.[b] He initiated a trade war with China, withdrew the U.S. from several international agreements,[c] and met with North Korean leader Kim Jong Un without progress on denuclearization. He responded to the COVID-19 pandemic with the CARES Act,[d] and downplayed its severity. He was impeached in 2019 for abuse of power and obstruction of Congress, and in 2021 for incitement of insurrection; the Senate acquitted him in both cases.')" \
--together --batch-size=500 --threads=1 --total=500000 --init="drop table if exists test2; create table test2(id bigint, value text)" --load="insert into test2(value) values('Donald John Trump (born June 14, 1946) is an American politician, media personality, and businessman who served as the 45th president of the United States from 2017 to 2021. He won the 2024 presidential election as the nominee of the Republican Party and is scheduled to be inaugurated as the 47th president on January 20, 2025. Trump graduated with a bachelors degree in economics from the University of Pennsylvania in 1968. After becoming president of the family real estate business in 1971, he renamed it the Trump Organization. After a series of bankruptcies in the 1990s, he launched side ventures, mostly licensing the Trump name. From 2004 to 2015, he produced and hosted the reality television series The Apprentice. In 2015 he launched a presidential campaign. Trump won the 2016 presidential election. His election and policies sparked numerous protests. In his first term, he ordered a travel ban targeting Muslims and refugees, funded the Trump wall expanding the U.S.–Mexico border wall, and implemented a family separation policy at the border. He rolled back more than 100 environmental policies and regulations, signed the Tax Cuts and Jobs Act of 2017,[a] and appointed three justices to the Supreme Court.[b] He initiated a trade war with China, withdrew the U.S. from several international agreements,[c] and met with North Korean leader Kim Jong Un without progress on denuclearization. He responded to the COVID-19 pandemic with the CARES Act,[d] and downplayed its severity. He was impeached in 2019 for abuse of power and obstruction of Congress, and in 2021 for incitement of insurrection; the Senate acquitted him in both cases.')"

```

**Comment by sanikolaev** [comment]
> Let's compare that with manual sharding, e.g. here we do the same like with shards=2, but by writing to the shards manually:

json:
```
root@dev:/workdir# ./test.sh "curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk" 1000 12
Testing with following parameters:
Curl command: curl -H 'Content-type: application/x-ndjson' --data-binary @bulk.json http://localhost:9308/_bulk
Number of requests: 1000
Concurrency: 12
----------------------------------------
Single-thread curl test (sequential):
----------------------------------------
Progress: 1000/1000
Total time: 62.058742914s
Average time per request: 62.058ms

Concurrent test using parallel execution:
----------------------------------------
Total time: 11.195777086s
Average time per request: 11.195ms
root@dev:/workdir#
```

sql:
```

Threads ; Batch     ; Time        ; Total Docs  ; Docs/Sec    ; Avg QPS     ; p99 QPS     ; p95 QPS     ; p5 QPS      ; p1 QPS      ; Lat Avg     ; Lat p50     ; Lat p95     ; Lat p99     ;
1       ; 500       ; 00:53       ; 500000      ; 9402        ; 19          ; 20          ; 19          ; 18          ; 18          ; 53.2        ; 52.5        ; 56.5        ; 64.5        ;
```

**Comment by donhardman** [comment]
Unblocked

**Comment by PavelShilin89** [comment]
It's probably a bug, `_bulk` doesn't support `insert` and requires `_index`, but using `_index` also gives an error, also we recently made support for `table` instead of `index`, which could also be a problem. I have tested all options, including the combination, `insert, index, delete table...`.
The order of queries and field types probably also plays a role. I could not find a working variant for my table, with different field types. 
```                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
~ mysql -h0 -P9306 -e ‘DROP TABLE IF EXISTS t;’
~ mysql -h0 -P9306 -e ‘CREATE TABLE t (id BIGINT, text TEXT) shards=3 rf=1;’
~ mysql -h0 -P9306 -e ‘INSERT INTO t (id, text) VALUES (1, “A”);’
 ~ mysql -h0 -P9306 -e ‘SELECT * FROM t ORDER BY id ASC;’
+------+------+
| id | text |
+------+------+
| 1 | A |
+------+------+

~ echo -e ‘{’insert‘:{’table‘:’t‘,’id‘:2}}\n{’text‘:’B‘}\n{’update‘:{’table‘:’t‘,’id‘:1}}\n{’doc‘:{’text‘:’C‘}}\n{’delete‘:{’table‘:’t‘,’id‘:2}}’ > bulk.json

~ cat bulk.json
{‘insert’:{‘table’:‘t’,‘id’:2}}
{‘text’: ‘B’}
{‘update’:{‘table’:‘t’,‘id’:1}}
{‘doc’:{‘text’: ‘C’}}}
{‘delete’:{‘table’:‘t’,‘id’:2}}

~ curl -s -H ‘Content-type: application/x-ndjson’ --data-binary @bulk.json http://localhost:9308/_bulk; echo
{‘error’:{‘type’: ‘action_request_validation_exception’, ‘reason’:‘\’_index\‘ property missing’}, ‘status’:400}
```

**Comment by donhardman** [comment]
`_bulk` is the Elasticsearch endpoint, and Elasticsearch does not have a "table" field. Instead, we should use "index" and follow Elasticsearch syntax. The "table" parameter should not work here.

Please refer to the original documentation of this proxied endpoint designed to work with Elasticsearch-like queries: https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html

**Comment by PavelShilin89** [comment]
@donhardman Already implemented test covering cases:

Insertions only
Deletion only
Only update
And multiple queries but identical. 

However, when testing mixed operations, an error occurs: 
**Created table:**
```
mysql -h0 -P9306 -e «create table test2 (id BIGINT, model TEXT, storage_capacity INTEGER, color string, release_year INTEGER, price FLOAT, discounted_price FLOAT, sold BOOL, date_added TIMESTAMP, product_codes MULTI, values MULTI64, additional_info JSON, vector float_vector knn_type='hnsw' knn_dims='4' hnsw_similarity='l2') shards=3 rf=1; »; echo $?
```
**Documents inserted**
```
~ mysql -h0 -P9306 -e "SELECT * FROM test2 order by id asc;"

+------+---------------+------------------+-------+--------------+-------------+------------------+------+------------+---------------+---------------------------------------+------------------------------------------------+-------------------------------------+
| id   | model         | storage_capacity | color | release_year | price       | discounted_price | sold | date_added | product_codes | values                                | additional_info                                | vector                              |
+------+---------------+------------------+-------+--------------+-------------+------------------+------+------------+---------------+---------------------------------------+------------------------------------------------+-------------------------------------+
|    1 | Updated Model |              512 | black |         2022 | 1399.989990 |      1299.989990 |    0 | 2661794816 | 10,11,12      | 123456789123456789,987654321987654321 | {"features":["New feature 1","New feature 2"]} | 0.700000,0.800000,0.900000,1.000000 |
|    4 | iPhone 14     |              256 | black |         2022 |  999.989990 |       899.989990 |    1 | 1661990400 | 19,20,21      | 1234567890123456789                   | {"features":["A16 Bionic","Dynamic Island"]}   | 0.100000,0.200000,0.300000,0.400000 |
|    5 | Pixel 7       |              128 | white |         2022 |  699.989990 |       649.989990 |    0 | 1661990400 | 16,17,18      | 223344556677889900,998877665544332211 | {"features":["Tensor G2","120Hz display"]}     | 0.400000,0.500000,0.600000,0.700000 |
|    6 | OnePlus 10    |              256 | blue  |         2022 |  799.989990 |       749.989990 |    0 | 1661990400 | 22,23,24      | 111222333444555666,777888999000111222 | {"features":["Snapdragon 8","OxygenOS"]}       | 0.200000,0.300000,0.400000,0.500000 |
|    7 | Xiaomi 12     |              128 | gray  |         2022 |  699.989990 |       649.989990 |    1 | 1661990400 | 25,26,27      | 333444555666777888,999000111222333444 | {"features":["MIUI 13","5G"]}                  | 0.300000,0.400000,0.500000,0.600000 |
+------+---------------+------------------+-------+--------------+-------------+------------------+------+------------+---------------+---------------------------------------+------------------------------------------------+-------------------------------------+
```

**1 example** 
```
~ cat -A bulk_operations.json

{“delete”:{“_index”:“test2”,“_id”:1}}$
{“index”:{“_index”:“test2”,“_id”:1}}$
{“model”:“C”,“storage_capacity”:256,“color”:“black”,“release_year”:2023,“price”:799.99,“discounted_price”:749. 99,“sold”:0,“date_added”:1672531200,“product_codes”:[10,11,12],“values”:[1234567890123456789,9876543210987654321],“additional_info”:{“features”:[“Feature A”,“Feature B”]},“vector”:[0.5,0.6,0.7,0.8]}$
{“index”:{“_index”:“test2”,“_id”:2}}$
{“model”:“Updated B”,“storage_capacity”:128,“color”:“blue”,“release_year”:2024,“price”:899.99,“discounted_price”:849. 99,“sold”:1,“date_added”:1672531200,“product_codes”:[40,41,42],“values”:[222222222222222222,333333333333333333],“additional_info”:{“features”:[“Feature X”,“Feature Y”]},“vector”:[0.1,0.2,0.3,0.4]}$
{“update”:{“_index”:“test2”,“_id”:6}}$
{“doc”:{“price”:749.99,“discounted_price”:699.99}}$

~ curl -s -H 'Content-type: application/x-ndjson' --data-binary @bulk_operations.json http://localhost:9308/_bulk; echo $?

{“error”: “Something went wrong”}0
```
**1 example** 
```
~ cat -A bulk_operations.json

{"index":{"_index":"test2","_id":10}}$
{"model":"New Model","storage_capacity":128,"color":"red","release_year":2023,"price":799.99,"discounted_price":749.99,"sold":0,"date_added":1672531200,"product_codes":[30,31,32],"values":[1234567890123456789,9876543210987654321],"additional_info":{"features":["Feature A","Feature B"]},"vector":[0.5,0.6,0.7,0.8]}$
{"index":{"_index":"test2","_id":8}}$
{"model":"Replaced Model","storage_capacity":256,"color":"blue","release_year":2024,"price":899.99,"discounted_price":849.99,"sold":1,"date_added":1672531200,"product_codes":[40,41,42],"values":[222222222222222222,333333333333333333],"additional_info":{"features":["Feature X","Feature Y"]},"vector":[0.1,0.2,0.3,0.4]}$
{"update":{"_index":"test2","_id":6}}$
{"doc":{"price":749.99,"discounted_price":699.99}}$
{"delete":{"_index":"test2","_id":7}}$

~ curl -s -H 'Content-type: application/x-ndjson' --data-binary @bulk_operations.json http://localhost:9308/_bulk; echo $?

{"error":"Undefined array key \"index\""}0
```

**Comment by donhardman** [comment]
@PavelShilin89 I have implemented files for mixed bulk operation, let's test again

**Comment by PavelShilin89** [comment]
Blocked by https://github.com/manticoresoftware/manticoresearch/issues/3079

**Comment by donhardman** [comment]
Unblocked, let's continue to test with the latest kit.

**Comment by PavelShilin89** [comment]
@donhardman Testing performed in PR - https://github.com/manticoresoftware/manticoresearch/pull/3073. The test passes correctly. Request approval to merge.

**Comment by marclaporte** [comment]
@PavelShilin89 I still see one failed test.

**Comment by PavelShilin89** [comment]
@donhardman `test-distributed-inserts.rec` - works correctly. 
When creating the same test but already with a sharded table in the cluster I got an error that curl is trying to access directly to the shard name and not the table itself, as well as a strange behavior that the cluster hash is specified in the error and not the given name.  
```
~ curl -s -X POST http://localhost:1308/insert -d '{"cluster": "replication", "table": "test2", "id": 1, "doc": {"model": "iPhone 13 Pro", "storage_capacity": 256, "color": "silver", "release_year": 2021, "price": 1099.99, "discounted_price": 989.99, "sold": 1, "date_added": 1591362342000, "product_codes": [1,2,3], "values": [523456764345678976,98765409877866654098,1109876543450987650987], "additional_info": {"features": ["ProMotion display", "A15 Bionic chip", "Ceramic Shield front cover"]}, "vector": [0.773448,0.312478,0.137971,0.459821]}}'; echo

{"error":{"type":"parse_exception","reason":"wrong table at cluster syntax, use \"cluster\": \"replication\" and \"index\": \"system.test2_s2\" properties, instead of 'replication:system.test2_s2'","table":"replication:system.test2_s2","index":"replication:test2"},"status":400}

~ curl -s -X POST http://localhost:1308/insert -d '{"table": "test2", "id": 2, "doc": {"model": "Galaxy S21 Ultra", "storage_capacity": 128, "color": "black", "release_year": 2021, "price": 1199.99, "discounted_price": 1099.99, "sold": 1, "date_added": 1609459200000, "product_codes": [4,5,6], "values": [1234567890123456789,9876543210987654321], "additional_info": {"features": ["Dynamic AMOLED 2X", "Exynos 2100", "108MP camera"]}, "vector": [0.5,0.4,0.3,0.2]}}'; echo

{"error":{"type":"action_request_validation_exception","reason":"table 'system.test2_s2' is a part of cluster 'e3d112439806feecd6da4d4cb4544210', use \"cluster\":\"e3d112439806feecd6da4d4cb4544210\" and \"table\":\"system.test2_s2\" properties","table":"system.test2_s2","index":"test2"},"status":409}

**Comment by donhardman** [comment]
I left some comments in the pull request. Please review and adjust tests if needed.

**Comment by PavelShilin89** [comment]
Merged to master
